File: WangLulu/DP/WordBreak.java
package DP;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class WordBreak {
        public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        boolean [] dp = new boolean [n+1];
        Arrays.fill(dp,false);
        Set<String> set = new HashSet<>();
        for(String word : wordDict){
            set.add(word);
        } 
        dp[0] = true;
        for(int i = 1;i<=n;i++){
            for(int j = 0;j<i;j++){
                String seg = s.substring(j,i);
                if(dp[j]&&set.contains(seg)){
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }
}

File: ShengJie Wang/Practice_13/WordBreak.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp(s.size() + 1, false);
        dp[0] = true;
        for (int i = 0; i < s.size(); i++) {
            if (!dp[i]) {
                continue;
            }
            for (auto& word : wordDict) {
                int match_to = i + word.size();
                if (match_to <= s.size()) {
                    dp[match_to] = dp[match_to] || match(s, i, word);
                }
            }
            
        }
        return dp[s.size()];
    }

    bool match(string& s, int start, string& word) {
        if (start + word.size() - 1 > s.size()) {
            return false;
        }

        for(int i = 0; i < word.size(); i++) {
            if(s[start + i] != word[i]) {
                return false;
            }
        }
        return true;
    }

};